packages/shared/src/error.ts

```diff
@@ -0,0 +1,80 @@
+const ERROR_CODES = {
+  BAD_REQUEST: 400,
+  NOT_FOUND: 404,
+
+  INTERNAL_SERVER_ERROR: 500,
+  INTERNAL_CLIENT_ERROR: 500,
+  URL_GENERATION_FAILED: 500,
+  UPLOAD_FAILED: 500,
+  MISSING_ENV: 500,
+  FILE_LIMIT_EXCEEDED: 500,
+} as const;
+
+type ErrorCode = keyof typeof ERROR_CODES;
+
+function messageFromUnknown(cause: unknown, fallback?: string) {
+  if (typeof cause === "string") {
+    return cause;
+  }
+  if (cause instanceof Error) {
+    return cause.message;
+  }
+  if (
+    cause &&
+    typeof cause === "object" &&
+    "message" in cause &&
+    typeof cause.message === "string"
+  ) {
+    return cause.message;
+  }
+  return fallback ?? "An unknown error occurred";
+}
+
+export class UploadThingError extends Error {
+  public readonly cause?: Error;
+  public readonly code: ErrorCode;
+
+  constructor(opts: {
+    code: keyof typeof ERROR_CODES;
+    message?: string;
+    cause?: unknown;
+  }) {
+    const message = opts.message ?? messageFromUnknown(opts.cause, opts.code);
+
+    super(message);
+    this.code = opts.code;
+
+    if (opts.cause instanceof Error) {
+      this.cause = opts.cause;
+    } else if (opts.cause instanceof Response) {
```

Comment:
How do we feel about

```ts
else if (opts.cause instanceof Response) {
      this.cause = new Error(`Response ${opts.cause.status} ${opts.cause.statusText}`);
}
```

Reply:
Yea.

Alternatively we'd just have to parse the response before creating the error in the first place, so that we could actually use the error message from the server.

This is just for the `cause` though. When the response is the error, we have the `UploadthingError.fromResponse()` we can use